pkg/raft/storage.go

```diff
@@ -73,25 +66,41 @@ type Storage interface {
 	// encountered an unavailable entry in [lo, hi).
 	Entries(lo, hi, maxSize uint64) ([]pb.Entry, error)
 
-	// Term returns the term of entry i, which must be in the range
-	// [FirstIndex()-1, LastIndex()]. The term of the entry before
-	// FirstIndex is retained for matching purposes even though the
-	// rest of that entry may not be available.
+	// Term returns the term of the entry at the given index, which must be in the
+	// valid range: [FirstIndex()-1, LastIndex()]. The term of the entry before
+	// FirstIndex is retained for matching purposes even though the rest of that
+	// entry may not be available.
 	Term(i uint64) (uint64, error)
 	// LastIndex returns the index of the last entry in the log.
 	LastIndex() (uint64, error)
-	// FirstIndex returns the index of the first log entry that is
-	// possibly available via Entries (older entries have been incorporated
-	// into the latest Snapshot; if storage only contains the dummy entry the
-	// first log entry is not available).
+	// FirstIndex returns the index of the first log entry that is possibly
+	// available via Entries (older entries have been incorporated into the
+	// StateStorage.Snapshot; if storage only contains the dummy entry or initial
+	// snapshot then the first log entry is not available).
 	FirstIndex() (uint64, error)
+}
+
+// StateStorage provides read access to the state machine storage.
+type StateStorage interface {
 	// Snapshot returns the most recent snapshot.
-	// If snapshot is temporarily unavailable, it should return ErrSnapshotTemporarilyUnavailable,
-	// so raft state machine could know that Storage needs some time to prepare
-	// snapshot and call Snapshot later.
+	//
+	// Returns ErrSnapshotTemporarilyUnavailable if the snapshot is temporarily
+	// unavailable. This indicates that StateStorage needs some time to prepare
+	// the snapshot, so raft tries to access it later.
 	Snapshot() (pb.Snapshot, error)
 }
 
+// Storage is an interface that should be implemented by the application to
+// provide raft with access to the log and state machine storage.
+//
+// If any Storage method returns an error, the raft instance will become
+// inoperable and refuse to participate in elections; the application is
+// responsible for cleanup and recovery in this case.
```

Comment:
I don't know how much of this is true. Sounds like replica cordoning, but we don't have it AFAIK. The comments might need another revisit and some more detailed specs.

Reply:
Yeah, the "refuse to participate" part is definitely not something implemented in any way. I would change this to just say that in general storage errors are not handled gracefully.

Reply:
Done.